### PR TITLE
feat(hatchery/local): download worker from CDS API

### PR DESCRIPTION
**Pull request by yesnault**
Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

File: docs/content/docs/components/hatchery/local.md [hunk 1]
```diff
@@ -12,10 +12,10 @@ can launch CDS Worker on your host.
 
 Two prerequisites: 
 
-* your host must reach your CDS API on HTTP port or GRPC Port defined on your [CDS Configuration]({{< relref "/hosting/configuration.md">}})
-* you need the CDS Worker binary on your host. 
+* your host must reach your CDS API on HTTP port defined on your [CDS Configuration]({{< relref "/hosting/configuration.md">}})
+* the basedir (default /tmp) must allow to execute binaries
```
Comment (fsamin) [inline docs/content/docs/components/hatchery/local.md]: having /tmp with executable right is a security risk

File: engine/hatchery/local/local.go [hunk 2]
```diff
@@ -135,12 +132,27 @@ func (h *HatcheryLocal) Serve(ctx context.Context) error {
 		return fmt.Errorf("Cannot check local capabilities: %v", err)
 	}
 
+	h.BasedirDedicated = fmt.Sprintf("%s/%s", strings.TrimRight(strings.TrimSpace(h.Config.Basedir), "/"), h.Name)
```
Comment (fsamin) [inline engine/hatchery/local/local.go]: could you use https://golang.org/pkg/path/filepath/#Dir ?

File: docker-compose.yml [hunk 3]
```diff
@@ -105,6 +105,7 @@ services:
       CDS_HATCHERY_LOCAL_COMMONCONFIGURATION_URL: http://cds-hatchery-local:8086
       CDS_HATCHERY_LOCAL_COMMONCONFIGURATION_API_HTTP_URL: http://cds-api:8081
       CDS_HATCHERY_LOCAL_NBPROVISION: 5
+      CDS_HATCHERY_LOCAL_BASEDIR: /tmp
```
Comment (fsamin) [inline docker-compose.yml]: update ?

File: docs/content/docs/components/hatchery/local.md [hunk 4]
```diff
@@ -12,10 +12,14 @@ can launch CDS Worker on your host.
 
 Two prerequisites: 
 
-* your host must reach your CDS API on HTTP port or GRPC Port defined on your [CDS Configuration]({{< relref "/hosting/configuration.md">}})
-* you need the CDS Worker binary on your host. 
+* your host must reach your CDS API on HTTP port defined on your [CDS Configuration]({{< relref "/hosting/configuration.md">}})
+<<<<<<< HEAD
```
Comment (fsamin) [inline docs/content/docs/components/hatchery/local.md]: this looks like a failed merge

File: engine/hatchery/local/local.go (hunk 2 above)
Comment (yesnault) [reply engine/hatchery/local/local.go]: yep, done

File: engine/hatchery/local/local.go [hunk 5]
```diff
@@ -135,12 +133,27 @@ func (h *HatcheryLocal) Serve(ctx context.Context) error {
 		return fmt.Errorf("Cannot check local capabilities: %v", err)
 	}
 
+	h.BasedirDedicated = filepath.Dir(fmt.Sprintf("%s/%s/", h.Config.Basedir, h.Name))
```
Comment (fsamin) [inline engine/hatchery/local/local.go]: filepath.Join

File: engine/hatchery/local/local.go [hunk 6]
```diff
@@ -151,6 +164,47 @@ func (h *HatcheryLocal) Serve(ctx context.Context) error {
...
+			return sdk.WrapError(err, "error while removing existing worker binary %s", workerFullPath)
+		}
+	}
+
+	log.Debug("copy worker binary into %s", workerFullPath)
+	fp, err := os.OpenFile(workerFullPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0700)
+	if err != nil {
+		return sdk.WithStack(err)
+	}
+	defer fp.Close()
```
Comment (fsamin) [inline engine/hatchery/local/local.go]: you should check this error because it flushes the file - want it could hide a useful error

Comment (yesnault) [reply engine/hatchery/local/local.go]: done

File: engine/hatchery/local/local.go (hunk 5 above)
Comment (yesnault) [reply engine/hatchery/local/local.go]: done

**Pull request merged**